transifex/native/parsing.py

```diff
@@ -304,86 +344,124 @@ def visit_Call(self, node):
-                    self.source_strings.append(source_string)
-                except Exception as e:
-                    raise AttributeError(
-                        'Invalid module/function format on line {} col {}: {}'.format(
-                            node.lineno, node.col_offset, e
-                        )
-                    )
+        # add the node for later processing
+        for import_obj in self.imports:
+            module_path = import_obj.module
+            func_name = import_obj.function
 
-    def _render_keyword(self, keyword):
-        """Render the given keyword to a proper value.
+            if module_path == current_module_path \
+                    and func_name == current_func_name:
+                self.function_calls.append(node)
 
-        Processes Keyword objects and returns the keyword name along with
-        a value of their corresponding type, i.e. a string, a number or a boolean.
+    def visit_BinOp(self, node):
+        """Store % operations to process later.
 
-        :param Keyword keyword: the keyword object
-        :return: a tuple like (<key>, <value>)
-        :rtype: tuple
+        We don't process them here because in _(...) % (...) statements
+        the _(...) node hasn't been processed yet when the modulo operation
+        is processed.
         """
-        if isinstance(keyword.value, ast.Str):
-            val = keyword.value.s
-        elif isinstance(keyword.value, ast.Num):
-            val = keyword.value.n
-        elif (isinstance(keyword.value, ast.Name)
-              and keyword.value.id in ('True', 'False')):
-            val = keyword.value.id == 'True'
+        if isinstance(node.op, ast.Mod) and isinstance(node.left, ast.Call):
+            func_call = node.left
+            self.visit_Call(func_call)
+            if self.function_calls and self.function_calls[-1] == func_call:
+                self.modulos[func_call] = node
         else:
-            val = None
+            self.generic_visit(node)
+
+
+def parse_source_strings(nodes):
+    """Parse the given function call nodes and return a list of SourceString
+    objects.
+
+    :param list nodes: a list of Node objects
+    """
+    strings = []
+    for node in nodes:
+        try:
+            string = node.args[0].s
+            # Context could be passed as an argument, e.g. t('str', 'context')
+            context = node.args[1].s if len(node.args) > 1 else None
+
+            # Find all custom parameters, e.g. developer comments etc
+            params = {}
+            for keyword in node.keywords:
+                name, value = render_keyword(keyword)
+                if value is not None:
+                    params[name] = value
+
+            # If no context was found before, maybe it was passed as a kwarg
+            if context is None:
+                context = params.pop(KEY_CONTEXT, None)
+
+            strings.append(SourceString(string, context, **params))
+        except Exception as e:
```

Comment:
Hm, I'm not sure, but while developing I saw that dealing with ast was pretty tricky in many cases, so I feel more confident catching the exception here and raising a more descriptive one. `parse_source_strings` is very high level, so it's our chance to log a meaningful error message for users.

Reply:
I am good with that but to the best of my knowledge, parse is going to raise if there is a syntax error. In case of any other exceptions, we should reconsider the algorithm of ours.